### PR TITLE
enh(user): Create an API endpoint to update the user's theme

### PR DESCRIPTION
**Pull request by callapa**
## Description

Create an API endpoint to update the user's theme

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

## Checklist

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 1]
```diff
@@ -16718,4 +16718,16 @@ msgid "Login with"
 msgstr "Se connecter avec"
 
 msgid "Can't resolve username from login claim using configured regular expression"
-msgstr "Impossible de résoudre le nom d'utilisateur à partir de la déclaration de connexion en utilisant l'expression régulière configurée"
\ No newline at end of file
+msgstr "Impossible de résoudre le nom d'utilisateur à partir de la déclaration de connexion en utilisant l'expression régulière configurée"
+
+msgid "Error on updating a user"
+msgstr "Erreur lors de la mise à jour d'un utilisateur (Raison : %s)"
```
Comment (kduret) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:16724]: `Reason` is not in english

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 2]
```diff
@@ -16718,4 +16718,16 @@ msgid "Login with"
...
+
+msgid "Error in reading the themes available to the user"
+msgstr "Erreur dans la lecture des thèmes disponibles pour l'utilisateur (Raison : %s)"
```
Comment (kduret) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:16727]: `Reason` is not in english

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 3]
```diff
@@ -16718,4 +16718,16 @@ msgid "Login with"
...
+
+msgid "Error when updating the user\'s theme"
+msgstr "Erreur lors de la mise à jour du thème de l'utilisateur (Raison : %s)"
```
Comment (kduret) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:16730]: `Reason` is not in english

File: tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+});
+
+it('tests the error message when user is not found', function () {
+    $this->readUserRepository
+        ->expects($this->once())
+        ->method('findUserById')
+        ->willReturn(null);
+    $useCase = new PatchUser($this->readUserRepository, $this->writeUserRepository);
+    $useCase($this->request, $this->presenter);
+    $this->assertEquals(new NotFoundResponse('User'), $this->presenter->getResponseStatus());
```
Comment (kduret) [inline tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php:52]:
```suggestion
    expect($this->presenter->getResponseStatus())->toBe(new NotFoundResponse('User'));
```

File: tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+});
+
+it('tests the exception while searching for the user', function () {
+    $this->readUserRepository
+        ->expects($this->once())
+        ->method('findUserById')
+        ->willThrowException(new UserException());
+    $useCase = new PatchUser($this->readUserRepository, $this->writeUserRepository);
+    $useCase($this->request, $this->presenter);
+    $this->assertEquals(
```
Comment (kduret) [inline tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php:62]: can you use `expect`

File: tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        ->method('findUserById')
+        ->willReturn($user);
+
+    $this->readUserRepository
+        ->expects($this->once())
+        ->method('findAvailableThemes')
+        ->willReturn([]);
+    $useCase = new PatchUser($this->readUserRepository, $this->writeUserRepository);
+    $useCase($this->request, $this->presenter);
+    $this->assertEquals(
```
Comment (kduret) [inline tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php:81]: can you use `expect`

File: tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        ->expects($this->once())
+        ->method('findUserById')
+        ->willReturn($user);
+    $this->readUserRepository
+        ->expects($this->once())
+        ->method('findAvailableThemes')
+        ->willReturn(['blue', 'green']);
+    $useCase = new PatchUser($this->readUserRepository, $this->writeUserRepository);
+    $useCase($this->request, $this->presenter);
+    $this->assertEquals(
```
Comment (kduret) [inline tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php:99]: can you use `expect`

File: tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        ->expects($this->once())
+        ->method('findUserById')
+        ->willReturn($user);
+    $this->readUserRepository
+        ->expects($this->once())
+        ->method('findAvailableThemes')
+        ->willThrowException(new UserException());
+    $useCase = new PatchUser($this->readUserRepository, $this->writeUserRepository);
+    $useCase($this->request, $this->presenter);
+    $this->assertEquals(
```
Comment (kduret) [inline tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php:117]: can you use `expect`

File: tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    $user->setTheme($this->request->theme);
+    $this->writeUserRepository
+        ->expects($this->once())
+        ->method('updateUser')
+        ->with($user)
+        ->willThrowException(new UserException());
+
+    $useCase = new PatchUser($this->readUserRepository, $this->writeUserRepository);
+    $useCase($this->request, $this->presenter);
+    $this->assertEquals(
```
Comment (kduret) [inline tests/php/Core/Application/Configuration/User/UseCase/PatchUser/PatchUserTest.php:142]: can you use `expect`

File: src/Core/Application/Security/User/Repository/WriteUserRepositoryInterface.php [hunk 10]
```diff
@@ -23,7 +23,9 @@
 
 namespace Core\Application\Security\User\Repository;
 
+use Core\Domain\Configuration\User\Model\User as ConfigurationUser;
```
Comment (kduret) [inline src/Core/Application/Security/User/Repository/WriteUserRepositoryInterface.php:26]: seems not used

File: src/Core/Application/Security/User/Repository/WriteUserRepositoryInterface.php [hunk 11]
```diff
@@ -23,7 +23,9 @@
...
 use Core\Domain\Security\User\Model\User;
+use Core\Infrastructure\Security\User\Repository\DbWriteUserRepository;
```
Comment (kduret) [inline src/Core/Application/Security/User/Repository/WriteUserRepositoryInterface.php:28]:
```suggestion
```

File: src/Core/Application/Security/User/Repository/WriteUserRepositoryInterface.php (hunk 10 above)
Comment (callapa) [reply src/Core/Application/Security/User/Repository/WriteUserRepositoryInterface.php:26]: my bad

File: src/Core/Application/Security/User/Repository/WriteUserRepositoryInterface.php (hunk 11 above)
Comment (callapa) [reply src/Core/Application/Security/User/Repository/WriteUserRepositoryInterface.php:28]: my bad

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 12]
```diff
@@ -16718,4 +16718,16 @@ msgid "Login with"
...
 msgid "Can't resolve username from login claim using configured regular expression"
-msgstr "Impossible de résoudre le nom d'utilisateur à partir de la déclaration de connexion en utilisant l'expression régulière configurée"
\ No newline at end of file
+msgstr "Impossible de résoudre le nom d'utilisateur à partir de la déclaration de connexion en utilisant l'expression régulière configurée"
+
+msgid "Error on updating a user"
+msgstr "Erreur lors de la mise à jour d'un utilisateur"
+
+msgid "Error in reading the themes available to the user"
+msgstr "Erreur dans la lecture des thèmes disponibles pour l'utilisateur)"
```
Comment (kduret) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:16727]:
```suggestion
msgstr "Erreur dans la lecture des thèmes disponibles pour l'utilisateur"
```

File: src/Core/Application/Configuration/User/UseCase/PatchUser/PatchUser.php [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+     * @param ReadUserRepositoryInterface $readUserRepository
+     * @param WriteUserRepositoryInterface $writeUserRepository
+     */
+    public function __construct(
+        private ReadUserRepositoryInterface $readUserRepository,
+        private WriteUserRepositoryInterface $writeUserRepository
+    ) {
+    }
+
+    public function __invoke(PatchUserRequest $request, PatchUserPresenterInterface $presenter): void
```
Comment (kduret) [inline src/Core/Application/Configuration/User/UseCase/PatchUser/PatchUser.php:49]: missing phpdoc

**Review by kduret — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

File: src/Core/Infrastructure/Configuration/User/Api/PatchUser/PatchUserController.php [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+{
+    use LoggerTrait;
+
+    /**
+     * @param Request $request
+     * @param mixed $user
+     * @param PatchUser $useCase
+     * @param PatchUserPresenterInterface $presenter
+     * @return object
+     */
```
Comment (adr-mo) [inline src/Core/Infrastructure/Configuration/User/Api/PatchUser/PatchUserController.php:43]: Simple re-order of the arguments to match method signature
```suggestion
    /**
     * @param Request $request
     * @param PatchUser $useCase
     * @param mixed $user
     * @param PatchUserPresenterInterface $presenter
     * @return object
     */
```

**Review by kduret — APPROVED** [review]

**Review by adr-mo — APPROVED** [review]

**Pull request merged**